core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+//! handshake and connection logic.
+#![allow(clippy::significant_drop_in_scrutinee)]
+
+use core::marker::PhantomData;
+use std::{
+    collections::HashMap,
+    io::{Read, Write},
+    net::{Shutdown, TcpListener, TcpStream, ToSocketAddrs},
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use parking_lot::Mutex;
+
+use eyre::WrapErr as _;
+use iroha_crypto::{
+    ursa::{
+        encryption::symm::{prelude::ChaCha20Poly1305, Encryptor, SymmetricEncryptor},
+        kex::{x25519::X25519Sha256, KeyExchangeScheme},
+        keys::{PrivateKey as UrsaPrivateKey, PublicKey as UrsaPublicKey},
+        CryptoError,
+    },
+    PublicKey,
+};
+use iroha_logger::{debug, error, info, trace};
+use parity_scale_codec::{Decode, Encode};
+use rand::{Rng, RngCore};
+use thiserror::Error;
+
+use crate::{handler::ThreadHandler, NetworkMessage, NetworkMessage::*, PeerId};
+
+/// Errors used in [`crate`].
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Failed to read or write
+    #[error("Failed IO operation.")]
+    Io(#[from] std::io::Error),
+    /// Handshake lengths don't match
+    #[error("Handshake length mismatch {0}")]
+    HandshakeLength(usize),
+    /// Parity Scale codec error
+    #[error("Parity Scale codec error")]
+    ParityScale(#[from] parity_scale_codec::Error),
+    /// Failed to create keys
+    #[error("Failed to create session key")]
+    Keys(#[from] CryptographicError),
+    /// Failed to resolve to any socket address
+    #[error("Failed to resolve to any socket address")]
+    Resolve,
+}
+
+/// Error in the cryptographic processes.
+#[derive(Debug, Error)]
+pub enum CryptographicError {
+    /// Decryption failed
+    #[error("Decryption failed")]
+    Decrypt(aead::Error),
+    /// Encryption failed
+    #[error("Encryption failed")]
+    Encrypt(aead::Error),
+    /// Ursa Cryptography error
+    #[error("Ursa Cryptography error")]
+    Ursa(CryptoError),
+}
```

Comment:
When I originally added this type, it was used to encapsulate the `Cryptographer` type and make all associated functions return that type of error. Now it seems that we're better off merging this with the Error enum.